File: cli/node/src/state.ts
/**
 * Agent mode state — set by the root program option handler,
 * read by commands and branding functions.
 */

import contract from "./contract/payload_contract.json" with { type: "json" };

let _agentMode = false;
let _currentCommand = "";
let _pendingNotice = "";
let _traceEnabled = true;
let _traceEventId = "";

export interface TraceStep {
	label: string;
	durationMs?: number;
	status?: string;
	metadata?: Record<string, unknown>;
	_startTime?: number;
}

const _traces: TraceStep[] = [];
let _traceContract: Record<string, unknown> | null = null;

function loadTraceContract(): Record<string, unknown> {
	if (!_traceContract) {
		_traceContract = (contract as Record<string, unknown>).traceEpilogue as Record<string, unknown> || {};
	}
	return _traceContract;
}

export function isAgentMode(): boolean {
	return _agentMode;
}

export function setAgentMode(val: boolean): void {
	_agentMode = val;
}

export function getCurrentCommand(): string {
	return _currentCommand;
}

export function setCurrentCommand(name: string): void {
	_currentCommand = name;
}

/**
 * Stash a Mem0 backend notice (Agent Mode unclaimed reminder) for end-of-
 * command surfacing. Called from the platform backend after each response so
 * the notice prints once per command regardless of how many sub-requests
 * fired. Last-write-wins is fine — the message text is identical.
 */
export function captureNotice(notice: string | null | undefined): void {
	if (notice) _pendingNotice = notice;
}

export function takeNotice(): string {
	const msg = _pendingNotice;
	_pendingNotice = "";
	return msg;
}

// ---------------------------------------------------------------------------
// Trace epilogue
// ---------------------------------------------------------------------------

export function setTraceEnabled(enabled: boolean): void {
	_traceEnabled = enabled;
}

export function isTraceEnabled(): boolean {
	return _traceEnabled && !isAgentMode();
}

export function setTraceEventId(eventId: string): void {
	_traceEventId = eventId;
}

export function getTraceEventId(): string {
	return _traceEventId;
}

export function addTraceStep(label: string, { start = false }: { start?: boolean } = {}): TraceStep {
	const step: TraceStep = { label };
	if (start && isTraceEnabled()) {
		step._startTime = performance.now();
	}
	if (isTraceEnabled()) {
		_traces.push(step);
	}
	return step;
}

export function finishTraceStep(step: TraceStep, status?: string): void {
	if (step._startTime !== undefined) {
		step.durationMs = Math.round(performance.now() - step._startTime);
	}
	if (status !== undefined) {
		step.status = status;
	}
}

export function getTraceSteps(): TraceStep[] {
	return [..._traces];
}

export function clearTraceSteps(): void {
	_traces.length = 0;
	_traceEventId = "";
}

export function formatTraceEpilogue(): string | null {
	if (!isTraceEnabled()) return null;

	const contract = loadTraceContract();
	const steps = getTraceSteps();
	if (steps.length === 0) return null;

	const minDuration = (contract.minDurationMs as number) ?? 5;
	const filtered = steps.filter((s) => s.durationMs === undefined || s.durationMs >= minDuration);
	if (filtered.length === 0) return null;

	const headerLabel = (contract.headerLabel as string) ?? "STEP";
	const headerDuration = (contract.headerDuration as string) ?? "DURATION";
	const headerStatus = (contract.headerStatus as string) ?? "STATUS";

	const labelWidth = (contract.labelWidth as number) ?? 18;
	const durationWidth = (contract.durationWidth as number) ?? 10;
	const statusWidth = (contract.statusWidth as number) ?? 10;

	const statusColors: Record<string, string> = (contract.statusColors as Record<string, string>) ?? {
		success: "green",
		failed: "red",
		skipped: "yellow",
		pending: "dim",
	};
	const defaultStatus = (contract.defaultStatus as string) ?? "success";
	const totalLabel = (contract.totalLabel as string) ?? "Total";

	const colorMap: Record<string, (s: string) => string> = {
		green: (s: string) => `\x1b[32m${s}\x1b[0m`,
		red: (s: string) => `\x1b[31m${s}\x1b[0m`,
		yellow: (s: string) => `\x1b[33m${s}\x1b[0m`,
		dim: (s: string) => `\x1b[2m${s}\x1b[0m`,
	};

	const reset = "\x1b[0m";
	const accent = "\x1b[35m";
	const dim = "\x1b[2m";

	const lines: string[] = [];
	lines.push(
		`  ${accent}${headerLabel.padEnd(labelWidth)}${" "}${headerDuration.padStart(
			durationWidth,
		)}${" "}${headerStatus.padStart(statusWidth)}${reset}`,
	);

	let totalMs = 0;
	for (const step of filtered) {
		const status = step.status ?? defaultStatus;
		const colorName = statusColors[status] ?? "dim";
		const colorFn = colorMap[colorName] ?? colorMap.dim;

		let durationStr: string;
		if (step.durationMs !== undefined) {
			durationStr = `${step.durationMs.toString().padStart(6)} ms`;
			totalMs += step.durationMs;
		} else {
			durationStr = "     —";
		}

		const statusText = colorFn(status.toUpperCase().padStart(statusWidth));
		lines.push(
			`  ${dim}${step.label.padEnd(labelWidth)}${reset}${" "}${durationStr.padStart(
				durationWidth,
			)}${" "}${statusText}`,
		);
	}

	lines.push(
		`  ${accent}${totalLabel.padEnd(labelWidth)}${" "}${accent}${totalMs
			.toString()
			.padStart(6)} ms${reset}${" "}${"".padStart(statusWidth)}`,
	);

	const traceEventId = getTraceEventId();
	if (traceEventId) {
		const traceCommand = ((contract.traceCommand as string) ?? "mem0 trace {event_id}").replace(
			"{event_id}",
			traceEventId,
		);
		lines.push("");
		lines.push(`  ${dim}  For full trace: ${traceCommand}${reset}`);
	}

	return lines.join("\n") + "\n";
}

export function emitTraceEpilogue(): void {
	const epilogue = formatTraceEpilogue();
	if (epilogue) {
		process.stderr.write("\n" + epilogue);
	}
}

File: cli/python/src/mem0_cli/state.py
"""Agent mode state — set by the root callback, read by commands and branding."""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


_agent_mode: bool = False
_current_command: str = ""
_pending_notice: str = ""
_traces: list["TraceStep"] = []
_trace_enabled: bool = True
_trace_event_id: str = ""
_trace_contract: dict[str, Any] | None = None


def _load_trace_contract() -> dict[str, Any]:
    global _trace_contract
    if _trace_contract is None:
        contract_path = Path(__file__).parent / "contract" / "payload_contract.json"
        with open(contract_path, encoding="utf-8") as f:
            contract = json.load(f)
        _trace_contract = contract.get("traceEpilogue", {})
    return _trace_contract


@dataclass
class TraceStep:
    """A single trace step captured during command execution."""

    label: str
    duration_ms: int | None = None
    status: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
    _start_time: float | None = field(default=None, repr=False, compare=False)

    def start(self) -> None:
        """Start timing this step."""
        self._start_time = time.perf_counter()

    def finish(self, status: str | None = None) -> None:
        """Finish timing this step and calculate duration."""
        if self._start_time is not None:
            self.duration_ms = int((time.perf_counter() - self._start_time) * 1000)
        if status is not None:
            self.status = status


def is_agent_mode() -> bool:
    return _agent_mode


def set_agent_mode(val: bool) -> None:
    global _agent_mode
    _agent_mode = val


def get_current_command() -> str:
    return _current_command


def set_current_command(name: str) -> None:
    global _current_command
    _current_command = name


def capture_notice(notice: str | None) -> None:
    """Stash a Mem0 backend notice for end-of-command surfacing.

    Called from the platform backend after each response so the notice can
    be printed once per command (regardless of how many sub-requests fired).
    Last-write-wins is fine — the message text is identical across requests.
    """
    global _pending_notice
    if notice:
        _pending_notice = notice


def take_notice() -> str:
    """Return and clear the pending notice."""
    global _pending_notice
    msg = _pending_notice
    _pending_notice = ""
    return msg


# ---------------------------------------------------------------------------
# Trace epilogue
# ---------------------------------------------------------------------------

def set_trace_enabled(enabled: bool) -> None:
    """Enable or disable trace collection."""
    global _trace_enabled
    _trace_enabled = enabled


def is_trace_enabled() -> bool:
    """Return True if trace collection is enabled."""
    return _trace_enabled and not is_agent_mode()


def set_trace_event_id(event_id: str) -> None:
    """Set the event ID for the current command's trace."""
    global _trace_event_id
    _trace_event_id = event_id


def get_trace_event_id() -> str:
    """Get the event ID for the current command's trace."""
    return _trace_event_id


def add_trace_step(label: str, *, start: bool = False) -> TraceStep:
    """Add a new trace step and optionally start timing it.

    Returns the TraceStep instance so caller can call .finish() on it.
    """
    step = TraceStep(label=label)
    if start and is_trace_enabled():
        step.start()
    if is_trace_enabled():
        _traces.append(step)
    return step


def get_trace_steps() -> list[TraceStep]:
    """Return all collected trace steps."""
    return list(_traces)


def clear_trace_steps() -> None:
    """Clear all collected trace steps (called at end of command)."""
    global _traces, _trace_event_id
    _traces = []
    _trace_event_id = ""


def format_trace_epilogue() -> str | None:
    """Format the collected trace steps as a trace epilogue table.

    Returns None if tracing is disabled or no steps were collected.
    """
    from mem0_cli.branding import ACCENT_COLOR, DIM_COLOR

    if not is_trace_enabled():
        return None

    contract = _load_trace_contract()
    steps = get_trace_steps()

    if not steps:
        return None

    min_duration = contract.get("minDurationMs", 5)
    filtered = [s for s in steps if s.duration_ms is None or s.duration_ms >= min_duration]

    if not filtered:
        return None

    header_label = contract.get("headerLabel", "STEP")
    header_duration = contract.get("headerDuration", "DURATION")
    header_status = contract.get("headerStatus", "STATUS")

    label_width = contract.get("labelWidth", 18)
    duration_width = contract.get("durationWidth", 10)
    status_width = contract.get("statusWidth", 10)

    status_colors = contract.get("statusColors", {
        "success": "green",
        "failed": "red",
        "skipped": "yellow",
        "pending": "dim",
    })
    default_status = contract.get("defaultStatus", "success")

    from rich.table import Table
    from rich.text import Text

    table = Table(
        show_header=True,
        header_style=ACCENT_COLOR,
        show_edge=False,
        show_lines=False,
        padding=0,
    )
    table.add_column(header_label, width=label_width, style=DIM_COLOR)
    table.add_column(header_duration, width=duration_width, justify="right")
    table.add_column(header_status, width=status_width, justify="right")

    total_ms = 0
    for step in filtered:
        status = step.status or default_status
        color_map = {
            "green": "green",
            "red": "red",
            "yellow": "yellow",
            "dim": DIM_COLOR,
            "success": "green",
            "failed": "red",
            "skipped": "yellow",
            "pending": DIM_COLOR,
        }
        status_style = color_map.get(status_colors.get(status, ""), DIM_COLOR)

        if step.duration_ms is not None:
            duration_str = f"{step.duration_ms:>6} ms"
            total_ms += step.duration_ms
        else:
            duration_str = "     —"

        status_text = Text(status.upper(), style=status_style)
        table.add_row(f"  {step.label}", duration_str, status_text)

    total_label = contract.get("totalLabel", "Total")
    table.add_row(
        f"  [{ACCENT_COLOR}]{total_label}[/{ACCENT_COLOR}]",
        f"[{ACCENT_COLOR}]{total_ms:>6} ms[/{ACCENT_COLOR}]",
        "",
    )

    from rich.console import Console
    import io

    buf = io.StringIO()
    console = Console(file=buf, force_terminal=False)
    console.print(table)

    trace_event_id = get_trace_event_id()
    if trace_event_id:
        trace_command = contract.get("traceCommand", "mem0 trace {event_id}")
        trace_command = trace_command.replace("{event_id}", trace_event_id)
        buf.write(f"\n  [{DIM_COLOR}]  For full trace: {trace_command}[/{DIM_COLOR}]\n")

    return buf.getvalue()


def emit_trace_epilogue(console) -> None:
    """Emit the trace epilogue to the given console if enabled."""
    epilogue = format_trace_epilogue()
    if epilogue:
        console.print(epilogue)
